File: scala/sgf-parsing/src/main/scala/Sgf.scala
import scala.util.parsing.combinator.RegexParsers

object Sgf extends RegexParsers {

  type Tree[A] = Node[A]
  type Forest[A] = List[Tree[A]]
  case class Node[A](rootLabel: A, subForest: Forest[A] = List())

  type SgfTree = Tree[SgfNode]

  type SgfNode = Map[String, List[String]]

  override val skipWhitespace = false

  def normalChar: Parser[String] = """[^\\\]]""".r

  def escapedNewline: Parser[String] = """\\\n""".r ^^^ ""
  def whitespace: Parser[String] = """\s""".r ^^^ " "
  def escapedChar: Parser[String] = "\\" ~> """.""".r

  def cValueType = rep1(escapedNewline | escapedChar | whitespace | normalChar) ^^ (x => x.mkString)

  def propValue = '[' ~> cValueType <~ ']'

  def propIdent = """[A-Z]""".r
  def property = propIdent ~ rep1(propValue) ^^ {
    case ident ~ values => Map((ident, values))
  }

  def node: Parser[SgfNode] = ";" ~> opt(property) ^^ {
    case Some(x) => x
    case None => Map()
  }

  def tree: Parser[SgfTree] = "(" ~> rep1(node) ~ rep(tree) <~ ")" ^^ {
    case Nil ~ trees => Node(Map())
    case xs ~ trees => nodesToTree(xs, trees)
  }

  def nodesToTree(nodes: List[SgfNode], trees: List[SgfTree]): SgfTree =
    nodes match {
      case Nil => throw new IllegalStateException("Can only create tree from non-empty nodes list")
      case x::Nil => Node(x, trees)
      case x::xs => Node(x, List(nodesToTree(xs, trees)))
    }

  def parseSgf(text: String): Option[SgfTree] =
    parse(tree, text) match {
        case Success(expression, _) => Some(expression)
        case _ => None
      }
}

File: haskell/sgf-parsing/haskell/sgf-parsing/Sgf.hs
{-# LANGUAGE OverloadedStrings #-}

module Sgf (parseSgf) where

import Data.Map (Map)
import qualified Data.Map as Map
import Data.Char (isUpper, isSpace)
import Data.Tree
import Data.Text (Text, pack)
import Control.Applicative (some)
import Control.Monad (void)
import Data.Attoparsec.Text

parseSgf :: Text -> Maybe (Tree (Map Text [Text]))
parseSgf = maybeResult . parse treeParser

type SgfTree = Tree SgfNode
type SgfNode = Map Text [Text]

treeParser :: Parser SgfTree
treeParser = do
	void $ char '('
	nodes <- some nodeParser
	trees <- many' treeParser
	void $ char ')'
	return (nodesToTree nodes trees)

nodeParser :: Parser SgfNode
nodeParser = do
	void $ char ';'
	properties <- many' propertyParser
	return (Map.fromList properties)

propertyParser :: Parser (Text, [Text])
propertyParser = do
	key <- some (satisfy isUpper)
	values <- some valueParser
	return (pack key, values)

valueParser :: Parser Text
valueParser = do	
	void $ char '['
	value <- valueTextParser "" '['
	return value

valueTextParser :: String -> Char -> Parser Text
valueTextParser acc previous = do
	current <- anyChar
	case (current, previous) of 
		(']', '\\')  -> valueTextParser (']' : acc) current
		(']', _)     -> return (pack $ reverse acc)
		('\n', '\\') -> valueTextParser (acc) current
		('\n', _)    -> valueTextParser (' ' : acc) current
		('\\', '\\') -> valueTextParser ('\\' : acc) current
		('\\', _)    -> valueTextParser acc current
		_            -> if isSpace current then 
				            valueTextParser (' ' : acc) current
			            else 
			 	            valueTextParser (current : acc) current

nodesToTree :: [SgfNode] -> [SgfTree] -> SgfTree
nodesToTree [] _ = error "Can only create tree from non-empty nodes list"
nodesToTree [x] trees = Node x trees
nodesToTree (x:xs) trees = Node x [nodesToTree xs trees]